src/hotspot/os/linux/os_linux.cpp

```diff
@@ -911,6 +915,29 @@ bool os::create_thread(Thread* thread, ThreadType thr_type,
   // Calculate stack size if it's not specified by caller.
   size_t stack_size = os::Posix::get_initial_stack_size(thr_type, req_stack_size);
   size_t guard_size = os::Linux::default_guard_size(thr_type);
+
+  // If THPs are unconditionally enabled, the following scenario can lead to huge RSS:
+  // - parent thread spawns, in quick succession, multiple child threads
+  // - child threads are slow to start
+  // - thread stacks of future child threads are adjacent and get merged into one large VMA
+  //   by the kernel, and subsequently transformed into huge pages by khugepaged
+  // - child threads come up, place JVM guard pages, thus splinter the large VMA, splinter
+  //   the huge pages into many (still paged-in) small pages.
+  // The result of that sequence are thread stacks that are fully paged-in even though the
+  // threads did not even start yet.
+  // We prevent that by letting the glibc allocate a guard page, which causes a VMA with different
+  // permission bits to separate two ajacent thread stacks and therefore prevent merging stacks
+  // into one VMA.
+  if (PreventTHPsForThreadStacks) {
+    guard_size = MAX2(guard_size, os::vm_page_size());
```

Comment:
I think this would be better hidden in `os::Linux::default_guard_size` and only applied to JavaThreads.

Reply:
> I think this would be better hidden in os::Linux::default_guard_size

I don't like this, since it would keep those two mitigations we apply - adding the glibc guard page and altering the stack size to be THP unaligned - separate in code. I find it clearer to have them in one place.

> and only applied to JavaThreads.

No, it makes sense to apply this to any thread that does not by-default run with a glibc guard page. But in practice, it is as you wish it to be since only Java Threads run without glibc guard pages.